### DISCUSSION TITLE
Last Active Times

### DISCUSSION THREAD

**Discussion by GrizzlyAK** [zwave-js/zwave-js-ui#2287] [Ideas]
As far as I know, there is no way to _know_ in HA what the value of Last Active in the rightmost column of the CP is for a particular device. The `last_updated` and `last_changed` state object properties are lacking if you want to know if a battery powered device has checked in without it actually reporting a _change_.

I'm assuming that this Last Active includes these "check in" times, as well as any other communication with a device, including reports. I don't expect it will ever be state property of a device, but perhaps it could be retrieved with a service call to fetch the value from zwave-js? Maybe it's possible now, idk. But for checking if a battery device is dead, especially for someone like me who has them set to wake up every 2 hours, it would be extremely useful data. I've found that only a small percentage of my battery powered devices actually report battery levels correctly. I've had some reporting as high as 90%, even though completely dead and the device inoperable. Only one or two have EVER shown 0% when they quit working. It seems the only _real_ way to know is to see if it checks in.

**Comment by robertsLando** [comment]
@GrizzlyAK The last seen and last updated time in nodes and values are managed on z2m side (nont from zwavejs driver) so I dunno what you mean here as I dunno how hass handles them. About the wrong battery reports I dunno how this could be improved, maybe a priodic call to refreshValue or pollValue (with an high timeout) could be a possible fix but I let @AlCalzone answer to this

**Reply by AlCalzone** [reply]
Sounds like https://github.com/zwave-js/node-zwave-js/issues/2589 would solve a large part of this issue - although the refresh surely will be less frequent than 2 hours. That issue depends on https://github.com/zwave-js/node-zwave-js/issues/1608 however, which asks for a timestamp to go with each value.

**Reply by GrizzlyAK** [reply]
@robertsLando I don't think HA has access to these values from z2m, but didn't know if, through the integration somehow, that information could be passed to HA (some kind of new service call?). In my case, since I have my batt devs wake up every two hours, these values should be updated regularly (I presume). I assume that if I were to check them and saw that they hadn't updated in 3-4 hours, that the batt my be dead. That was my thinking on trying to solve my batt reporting issue. I don't really expect batt level reports to ever be very accurate and will likely vary widely between different devices/manufacturers. 

@AlCalzone the issues you posted look complex and may indeed solve this issue directly. At the moment, it _appears_ that all of my devices report battery level, just not accurately. However, that could be entirely dependent on how the device measures voltage. Most of mine use Lithium batteries (123) and I know that Lithiums don't exhibit the same life behavior as Alkalines in that they tend to retain voltage until the last minute and then die quickly. That may be a source of some of these reporting anomalies. Just guessing. 

I appreciate the response and information. Not a big issue for me, but was just looking for a solution  and wanted to ask and perhaps offer new ideas.

Let me know if I can do anything to test on the devices I have (zw and zw+). 

Thanks,
Shane

**Reply by jefft4** [reply]
@GrizzlyAK from my experience, a lot of battery sensors report a false battery level; it's either 100 or 0, nothing in between.  Mostly it's manufacturer-dependent (i.e., did they bother to measure the voltage or just to have a cheaper threshold switch circuit?)

@robertsLando - it would be very helpful to get zwave-js-ui's 'last active' timestamp in HA; it seems to be the only really correct 'when did this thing last transmit some data?' test.  Currently, I use the 'last_changed' timestamp of the node state entity to see if anything is possibly dead, but that's not entirely reliable for battery devices (e.g. some remotes don't seem to update it) and it's not informative for mains-powered nodes as their state is normally 'online' for a long time until the controller deems them possibly dead.   
Is that 'last active' attribute from the control panel passed through to the HA integration (via WS) in any way that we can try to access inside HA?

**Reply by AlCalzone** [reply]
> a lot of battery sensors report a false battery level; it's either 100 or 0, nothing in between.

I **think** this is two issues in one:
1. Many sensors don't regularly report the battery level. Since `node-zwave-js` [v10.12](https://github.com/zwave-js/node-zwave-js/blob/master/CHANGELOG.md#10120-2023-03-15) you should get updates once a week now.
2. It seems that the sensors don't actually report 0% for the lowest operating voltage (usually around 2.2 V for a 3.7 V battery), but scale linearly to 0 V. That way, an empty battery is actually around 60%, and if they manage to report "low battery", that gets translated to 0% by zwave-js.

**Reply by robertsLando** [reply]
> Is that 'last active' attribute from the control panel passed through to the HA integration (via WS) in any way that we can try to access inside HA?

I don't know, I just pass to the websocket the driver instance and nothing else so I dunno what kind information they pass to HA integration. This is a question to ask in HA core repo

**Reply by AlCalzone** [reply]
@robertsLando You're managing that attribute yourself in Z-UI I think?

**Reply by robertsLando** [reply]
Yeah

**Reply by robertsLando** [reply]
See: https://github.com/zwave-js/zwave-js-ui/blob/master/lib/ZwaveClient.ts#L4242

I update it when I receive a stats update from node that indicates I received a packet from it

**Reply by jefft4** [reply]
> > Is that 'last active' attribute from the control panel passed through to the HA integration (via WS) in any way that we can try to access inside HA?
> 
> I don't know, I just pass to the websocket the driver instance and nothing else so I dunno what kind information they pass to HA integration. This is a question to ask in HA core repo

I'll go and ask...!

**Reply by AlCalzone** [reply]
> Yeah

so it isn't exposed via zwave-js-server then.

**Reply by robertsLando** [reply]
Yeah dunno BTW if they do something similar on their side too
